src/xmerl_c14n.erl

```diff
@@ -0,0 +1,529 @@
+    c14n(Elem, true).
+
+%% @doc Puts an XML document or element into canonical form, as a string.
+%%
+%% If the Comments argument is true, preserves comments in the output.
+-spec c14n(XmlThing :: xml_thing(), Comments :: boolean()) -> string().
+c14n(Elem, Comments) ->
+    c14n(Elem, Comments, []).
+
+%% @doc Puts an XML document or element into canonical form, as a string.
+%%
+%% If the Comments argument is true, preserves comments in the output. Any
+%% namespace prefixes listed in InclusiveNs will be left as they are and not
+%% modified during canonicalization.
+-spec c14n(XmlThing :: xml_thing(), Comments :: boolean(), InclusiveNs :: [string()]) -> string().
+c14n(Elem, Comments, InclusiveNs) ->
+    {NakedNss, _} = naked_nss(Elem, #{}, 0),
+    logger:debug("NakedNss: ~p~n", [lists:keysort(2, maps:to_list(NakedNss))]),
+    SortedNss = [ Prefix || {Prefix, _} <-
+                               lists:keysort(2, maps:to_list(NakedNss)) ],
+    %% FIXME: This still won't work. "ns14" < "ns4" in erlang and c14n/6
+    %% will put prefixes in the wrong order.
+    %% But we are saved by Skatteverket's removal of prefixes.
+    %% So we will leave this for a better day when we want to support
+    %% the complete standard.
+    %% Also they should probably be sorted in url order, I think...
+    FirstTagNss = lists:usort(InclusiveNs ++ SortedNss),
+    lists:flatten(lists:reverse(c14n(Elem, [], [], Comments, FirstTagNss, []))).
+
+naked_nss(#xmlDocument{content = Kids}, NSMap, Seq) ->
+    naked_nss(Kids, NSMap, Seq);
+naked_nss(#xmlElement{content = Kids, nsinfo = {Prefix, _},
+                      attributes = Attrs, namespace = Ns}, NSMap1, Seq1) ->
+    case [ A || #xmlAttribute{nsinfo = {"xmlns", P}} = A <- Attrs, P == Prefix ] of
+        [#xmlAttribute{}] ->
+            %% ns defined here, so not needed
+            naked_nss(Kids, NSMap1, Seq1);
+        [] ->
+            case Ns#xmlNamespace.default of
+                [] ->
+                    %% No defaultns, so we need this
+                    {NSMap2, Seq2} = insert_ns([Prefix], NSMap1, Seq1),
+                    naked_nss(Kids, NSMap2, Seq2);
+                _ ->
+                    %% Element has defaultns, so not needed
+            naked_nss(Kids, NSMap1, Seq1)
+            end
+    end;
+naked_nss(#xmlElement{content = Kids}, NSMap, Seq) ->
+    naked_nss(Kids, NSMap, Seq);
+naked_nss(#xmlText{}, NSMap, Seq) ->
+    {NSMap, Seq};
+naked_nss(#xmlComment{}, NSMap, Seq) ->
+    {NSMap, Seq};
+naked_nss(#xmlPI{}, NSMap, Seq) ->
+    {NSMap, Seq};
+naked_nss([H | T], NSMap1, Seq1) ->
+    {NSMap2, Seq2} = naked_nss(H, NSMap1, Seq1),
+    naked_nss(T, NSMap2, Seq2);
+naked_nss([], NSMap, Seq) ->
+    {NSMap, Seq}.
+
+insert_ns([NS | T], NSMap, Seq) ->
+    case maps:is_key(NS, NSMap) of
+        true ->
+            insert_ns(T, NSMap, Seq);
+        false ->
+            insert_ns(T, NSMap#{NS => Seq}, Seq+1)
+    end;
+insert_ns([], NSMap, Seq) ->
+    {NSMap, Seq}.
```

Comment:
This doesn't need to be recursive if we never call it with more than one prefix, and we don't seem to do that.

Reply:
yeah, I changed the function that calls it, so I can simplify